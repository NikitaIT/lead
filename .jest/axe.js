require('jest-axe/extend-expect');
require('@testing-library/jest-dom');
require('@testing-library/react');
