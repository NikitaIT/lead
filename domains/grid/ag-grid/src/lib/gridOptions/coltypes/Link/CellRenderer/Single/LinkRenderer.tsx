import { assert, isFelsely } from '@lead/std';
import { LinkRendererProps, CellData, CellValue } from './ports';

export function LinkRenderer<
  TProps extends LinkRendererProps<TCellData, TCellValue, TProps>,
  TCellData extends CellData = CellData,
  TCellValue extends CellValue = CellValue
>(props: TProps): React.ReactElement | null {
  /** if href not defined it's incorrect usage */
  assert(
    props.href,
    /** this msg should be placed outside of component, and provided as config validator */
    `Please, set colDef.cellRendererParams.href = (params: LinkCellRendererParams) => params.value`
  );
  /** we can override formater for link value or value getter */
  const valueFormatted = isFelsely(props.valueFormatted)
    ? props.value
    : props.valueFormatted;
  const noValue = isFelsely(props.value);
  /**
   * we expect that link href id is value
   * if id not defined we can skip href formatter
   */
  const mayBeLink = noValue && props.href(props);

  const hideLink = !mayBeLink;

  return (
    // we should provide same interface for all components
    // eslint-disable-next-line react/jsx-no-useless-fragment
    <>
      {hideLink ? (
        valueFormatted
      ) : (
        /** we can override formater for link href */
        <XHtml.a href={mayBeLink} target="_blank" rel="noreferrer">
          {valueFormatted}
        </XHtml.a>
      )}
    </>
  );
}
