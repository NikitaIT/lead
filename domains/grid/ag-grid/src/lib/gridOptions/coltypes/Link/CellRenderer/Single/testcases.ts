import { LinkRendererProps, CellData, CellValue } from './ports';

const _ = <T extends LinkRendererProps<CellData, CellValue, unknown>>(
  x: T
): T => x;

export const testcases = {
  'valueFormatted should be visible': _({
    value: 'value',
    valueFormatted: 'valueFormatted',
    href(params) {
      return `#${params.value}`;
    },
  } as const),
  'value should be visible': _({
    value: 'value',
    valueFormatted: null,
    href(params) {
      return `#${params.value}`;
    },
  } as const),
  'link should be hidden': _({
    value: undefined,
    valueFormatted: 'value',
    href(params) {
      return `#${params.value}`;
    },
  } as const),
} as const;
