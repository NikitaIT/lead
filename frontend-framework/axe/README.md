# frontend-framework-axe

This library was generated with [Nx](https://nx.dev).

## Building

Run `nx build frontend-framework-axe` to build the library.

## Running unit tests

Run `nx test frontend-framework-axe` to execute the unit tests via [Jest](https://jestjs.io).
