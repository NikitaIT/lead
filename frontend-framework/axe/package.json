{
  "name": "@lead/axe",
  "version": "0.0.1",
  "type": "commonjs"
}
