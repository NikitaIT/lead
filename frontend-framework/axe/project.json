{
  "name": "frontend-framework-axe",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "frontend-framework/axe/src",
  "projectType": "library",
  "targets": {
    "build": {
      "executor": "@nrwl/js:tsc",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/./frontend-framework/axe",
        "main": "frontend-framework/axe/src/index.ts",
        "tsConfig": "frontend-framework/axe/tsconfig.lib.json",
        "assets": ["frontend-framework/axe/*.md"]
      }
    },
    "publish": {
      "executor": "nx:run-commands",
      "options": {
        "command": "node tools/scripts/publish.mjs frontend-framework-axe {args.ver} {args.tag}"
      },
      "dependsOn": ["build"]
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["frontend-framework/axe/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "frontend-framework/axe/jest.config.ts",
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "codeCoverage": true
        }
      }
    }
  },
  "tags": []
}
