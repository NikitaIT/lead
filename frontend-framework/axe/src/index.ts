export * from './lib/frontend-framework-axe';
