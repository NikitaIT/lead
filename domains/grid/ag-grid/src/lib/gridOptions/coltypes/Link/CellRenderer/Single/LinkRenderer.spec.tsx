import { render } from '@testing-library/react';

import { LinkRenderer } from './LinkRenderer';
import { axe } from '@lead/axe';
import { testcases } from './testcases';

describe('LinkRenderer', () => {
  it.each(Object.entries(testcases))(
    'should render successfully %s',
    async (k, v) => {
      expect(XHtml.a).toBe('a');

      const { baseElement } = render(<LinkRenderer {...v} />);
      expect(baseElement).toBeTruthy();
      await axe.componentHaveNoViolationsAsync(baseElement);
      expect(baseElement).toMatchSnapshot();
    }
  );
  it('a11n works', async () => {
    try {
      const { baseElement } = render(<a href="#"></a>);
      expect(baseElement).toBeTruthy();
      await axe.componentHaveNoViolationsAsync(baseElement);
    } catch (e) {
      expect(e).toBeTruthy();
    }
  });
});
