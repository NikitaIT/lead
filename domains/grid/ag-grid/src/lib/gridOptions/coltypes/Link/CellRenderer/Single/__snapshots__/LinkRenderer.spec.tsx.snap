// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkRenderer should render successfully link should be hidden 1`] = `
<body>
  <div>
    value
  </div>
</body>
`;

exports[`LinkRenderer should render successfully value should be visible 1`] = `
<body>
  <div>
    <a
      href="#value"
      rel="noreferrer"
      target="_blank"
    >
      value
    </a>
  </div>
</body>
`;

exports[`LinkRenderer should render successfully valueFormatted should be visible 1`] = `
<body>
  <div>
    <a
      href="#value"
      rel="noreferrer"
      target="_blank"
    >
      valueFormatted
    </a>
  </div>
</body>
`;
